src/cpu-template-helper/src/fingerprint/compare.rs

```diff
@@ -0,0 +1,150 @@
+
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+    /// Difference detected between source and target.
+    #[error("Difference detected between source and target:\n{0}")]
+    DiffDetected(String),
+    /// Failed to serialize/deserialize JSON.
+    #[error("Failed to serialize/deserialize JSON: {0}")]
+    Serde(#[from] serde_json::Error),
+}
+
+#[derive(Serialize)]
+struct Diff<'a, T: Serialize> {
+    name: String,
+    prev: &'a T,
+    curr: &'a T,
+}
+
+pub fn compare(
+    prev: Fingerprint,
+    curr: Fingerprint,
+    filters: Vec<FingerprintField>,
+) -> Result<(), Error> {
+    let compare = |field: &FingerprintField, val1, val2| -> Result<String, serde_json::Error> {
+        if val1 != val2 {
+            let diff = Diff {
+                name: format!("{field:#?}"),
+                prev: val1,
+                curr: val2,
+            };
+            serde_json::to_string_pretty(&diff)
+        } else {
+            Ok("".to_string())
+        }
+    };
+
+    let results = filters
+        .into_iter()
+        .map(|filter| {
+            match filter {
+                FingerprintField::firecracker_version => compare(
+                    &filter,
+                    &prev.firecracker_version,
+                    &curr.firecracker_version,
+                ),
+                FingerprintField::kernel_version => {
+                    compare(&filter, &prev.kernel_version, &curr.kernel_version)
+                }
+                FingerprintField::microcode_version => {
+                    compare(&filter, &prev.microcode_version, &curr.microcode_version)
+                }
+                FingerprintField::bios_version => {
+                    compare(&filter, &prev.bios_version, &curr.bios_version)
+                }
+                FingerprintField::bios_revision => {
+                    compare(&filter, &prev.bios_revision, &curr.bios_revision)
+                }
+                FingerprintField::guest_cpu_config => {
+                    if prev.guest_cpu_config != curr.guest_cpu_config {
+                        let cpu_configs =
+                            vec![prev.guest_cpu_config.clone(), curr.guest_cpu_config.clone()];
+
+                        // This `strip()` call always succeed since the number of inputs is two.
+                        let cpu_configs = crate::template::strip::strip(cpu_configs).unwrap();
+
+                        let diff = Diff {
+                            name: "guest_cpu_config".to_string(),
+                            prev: &cpu_configs[0],
+                            curr: &cpu_configs[1],
+                        };
+                        serde_json::to_string_pretty(&diff)
+                    } else {
+                        Ok("".to_string())
+                    }
+                }
+            }
+        })
+        .collect::<Result<Vec<_>, serde_json::Error>>()?
+        .into_iter()
+        .filter(|result| !result.is_empty())
```

Comment:
I totally forgot, but there is a `filter_map` method that you can use to remove second loop.
In cases where you return value it would be: `Some(serde_json::to_string_pretty(&diff))`
and in case where you return empty string return `None`.